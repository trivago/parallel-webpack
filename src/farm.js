import chalk from 'chalk';
import { info } from './logger';
import pluralize from 'pluralize';
import jestWorker from 'jest-worker';
import settle from 'p-settle';

export const startFarm = (config, configPath, options = {}, workers) => {
    config = Array.isArray(config) ? config : [config];
    const silent = !!options.json;

    if (!silent) {
        info(
            'Building %s %s',
            chalk.yellow(config.length),
            pluralize('target', config.length),
        );
    }

    const builds = config.map((c, i) =>
        workers.build(configPath, options, i, config.length),
    );

    if (options.bail) {
        return Promise.all(builds);
    } else {
        return settle(builds).then(results =>
            Promise.all(
                results.map(
                    result =>
                        result.isFulfilled
                            ? result.value
                            : Promise.reject(result.reason),
                ),
            ),
        );
    }
};

export const createFarm = farmOptions =>
    new jestWorker(require.resolve('./webpackWorker'), {
        forkOptions: {
            stdio: [0, 1, 2, 'ipc'],
        },
        ...farmOptions,
    });
