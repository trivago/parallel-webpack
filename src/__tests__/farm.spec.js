import jestWorker from 'jest-worker';
import { createFarm, startFarm } from '../farm';
import settle from 'p-settle';

jest.mock('jest-worker');
jest.mock('p-settle');

jestWorker.mockImplementation(() => ({
    isConstructorCalled: true,
}));
settle.mockImplementation((...args) =>
    require.requireActual('p-settle')(...args),
);
process.env.FORCE_COLOR = 0;
describe('Farm utils', () => {
    describe('createFarm', () => {
        it('should create webpackWorkers with passed options', () => {
            const farm = createFarm({ json: false });
            expect(jestWorker).toHaveBeenCalledWith(
                require.resolve('../webpackWorker'),
                {
                    forkOptions: { stdio: [0, 1, 2, 'ipc'] },
                    json: false,
                },
            );
            expect(farm).toEqual({ isConstructorCalled: true });
        });
    });
    describe('startFarm', () => {
        const configs = [{}, {}, {}];
        let webpackOptions;
        let farm;
        beforeEach(() => {
            jest.clearAllMocks();
            jest.spyOn(console, 'log').mockImplementationOnce(() => {});
            webpackOptions = { optimize: false, env: 'prod', color: false };
            farm = {
                build: jest.fn((...args) => args),
            };
        });
        it('should run workers', async () => {
            const result = await startFarm(
                configs,
                '/path/to/cfg',
                webpackOptions,
                farm,
            );
            expect(result).toMatchSnapshot();
        });
        it('should bail out', async () => {
            webpackOptions = {
                ...webpackOptions,
                bail: true,
                json: true,
            };
            const farm = {
                build: jest.fn(
                    (configPath, options, i, length) =>
                        i === 1
                            ? Promise.reject('Failed at index 1')
                            : Promise.resolve('resolved index' + i),
                ),
            };
            let result;
            try {
                result = await startFarm(
                    configs,
                    '/path/to/cfg',
                    webpackOptions,
                    farm,
                );
                fail('it should go to catch');
            } catch (err) {
                expect(err).toBe('Failed at index 1');
                expect(result).not.toBeDefined();
                expect(console.log.mock.calls).toMatchSnapshot();
            }
        });

        it('should settle all results', async () => {
            webpackOptions = {
                ...webpackOptions,
                bail: false,
            };
            const farm = {
                build: jest.fn((configPath, options, i, length) => {
                    if (i === 1) {
                        return Promise.reject('Failed at index 1');
                    } else {
                        return Promise.resolve('resolved index' + i);
                    }
                }),
            };

            let result;
            try {
                result = await startFarm(
                    configs,
                    '/path/to/cfg',
                    webpackOptions,
                    farm,
                );
            } catch (error) {
                expect(console.log.mock.calls).toMatchSnapshot();
                expect(error).toBe('Failed at index 1');
                expect(result).not.toBeDefined();
            }
        });
    });
});
