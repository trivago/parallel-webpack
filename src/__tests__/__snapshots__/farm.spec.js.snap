// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Farm utils startFarm should bail out 1`] = `Array []`;

exports[`Farm utils startFarm should run workers 1`] = `
Array [
  Array [
    "/path/to/cfg",
    Object {
      "color": false,
      "env": "prod",
      "optimize": false,
    },
    0,
    3,
  ],
  Array [
    "/path/to/cfg",
    Object {
      "color": false,
      "env": "prod",
      "optimize": false,
    },
    1,
    3,
  ],
  Array [
    "/path/to/cfg",
    Object {
      "color": false,
      "env": "prod",
      "optimize": false,
    },
    2,
    3,
  ],
]
`;

exports[`Farm utils startFarm should settle all results 1`] = `
Array [
  Array [
    "[34m[WEBPACK][39m Building %s %s",
    "[33m3[39m",
    "targets",
  ],
]
`;
